### PR TITLE
Add support for non-column key for equijoin when eliminating cross join to inner join

### LINKED ISSUE #4442: Add support for non-column key for equijoin when eliminating cross join to inner join
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
Currently datafusion support to rewrite cross join to inner join. For example:
```sql
select * from test0 as t0 cross join test1 as t1 where t0.a = t1.a;
select * from test0 as t0 cross join test1 as t1 where t0.b = t1.b;
...
```

But only when the join keys are column, so the following sql will not rewrite to `inner join`.
```sql
select * from test0 as t0 cross join test1 as t1 where t0.a + 1 = t1.a * 2;
select * from test0 as t0 cross join test1 as t1 where t0.a * 2 = t1.a - 1;
...
```

**Describe the solution you'd like**
Add support for non-column key for equijoin when eliminating cross join to inner join  

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by ygf11**
# Which issue does this PR close?

Closes #4442.
Part of #4389.

# Rationale for this change
* Currently `EliminateCrossJoin` will only support eliminating cross join when the join keys are columns. We should support non-column join keys.
* This pr is part of #4389. 

# What changes are included in this PR?

* Add logics to support non-column join key in `EliminateCrossJoin`.
* Fix duplicated column name when join keys contain cast(column as xxx).
* Add unit tests.

# Are these changes tested?
Yes, unit tests will test it.

# Are there any user-facing changes?

### CODE REVIEW

**Comment by ygf11** [comment]
cc @alamb @jackwener @mingmwang

**Comment by liukun4515** [comment]
@ygf11 thanks for your contribution, could you please add some IT in the `datafusion/core/tests/sql/joins.rs` file with your SQL like 
```
select * from test0 as t0 cross join test1 as t1 where t0.a + 1 = t1.a * 2;
select * from test0 as t0 cross join test1 as t1 where t0.a * 2 = t1.a - 1;
```

**Comment by ygf11** [comment]
> thanks for your contribution, could you please add some IT in the datafusion/core/tests/sql/joins.rs file with your SQL like

@liukun4515 Added.

File: datafusion/core/tests/sql/joins.rs [hunk 1]
```diff
@@ -2304,3 +2304,99 @@ async fn error_cross_join() -> Result<()> {
...
+        let msg = format!("Creating logical plan for '{}'", sql);
+        let plan = ctx
+            .create_logical_plan(&("explain ".to_owned() + sql))
+            .expect(&msg);
+        let state = ctx.state();
+        let plan = state.optimize(&plan)?;
+        let expected = vec![
+            "Explain [plan_type:Utf8, plan:Utf8]",
+            "  Projection: t1.t1_id, t2.t2_id, t1.t1_name [t1_id:UInt32;N, t2_id:UInt32;N, t1_name:Utf8;N]",
+            "    Projection: t1.t1_id, t1.t1_name, t2.t2_id [t1_id:UInt32;N, t1_name:Utf8;N, t2_id:UInt32;N]",
```
Comment (mingmwang) [inline datafusion/core/tests/sql/joins.rs:2373]: Looks like these two projections are not collapsed. This is unrelated to this PR. But we need to take a look at the `PushDownProjection` rule.

File: datafusion/expr/src/logical_plan/builder.rs [hunk 2]
```diff
@@ -1012,6 +1013,68 @@ pub fn table_scan(
...
+    let alias_join_keys: Vec<Expr> = join_keys
+        .iter()
+        .map(|key| {
+            // The display_name() of cast expression will ignore the cast info, and show the inner expression name.
+            // If we do not add alais, it will throw same field name error in the schema when adding projection.
+            // For example:
+            //    input scan : [a, b, c],
+            //    join keys: [cast(a as int)]
+            //
+            //  then a and cast(a as int) will use the same field name - `a` in projection schema.
```
Comment (mingmwang) [inline datafusion/expr/src/logical_plan/builder.rs:1031]: Why the `Cast` expressions ignore the cast info ? Looks like this is not the expected behavior. 
I think no matter the Cast is explicit or implicit, it should display the cast info.

https://github.com/apache/arrow-datafusion/blob/cedb05aedf3cea030bfa8774b8575d8f4806a1c8/datafusion/expr/src/expr.rs#L1108-L1115

@andygrove @liukun4515 @jackwener 
How do you think?

File: datafusion/core/tests/sql/joins.rs (hunk 1 above)
Comment (liukun4515) [reply datafusion/core/tests/sql/joins.rs:2373]: cc @jackwener 
maybe @jackwener  has fixed that in https://github.com/apache/arrow-datafusion/pull/4487

**Comment by mingmwang** [comment]
The PR itself LGTM.

File: datafusion/core/tests/sql/joins.rs (hunk 1 above)
Comment (jackwener) [reply datafusion/core/tests/sql/joins.rs:2373]: Look like we need rebase and look the new result.

**Review by jackwener — APPROVED** [review]
LGTM generally.
But there may be some problem cooperate with other rules, but not related to this one.
I will try to investigate and think about them.

**Review by alamb — APPROVED** [review]
I looked at several of the tests carefully and skimmed the rest of this PR -- I also appreciate @jackwener 's PR.

Thank you very much @ygf11  -- I'll plan to merge this PR tomorrow unless there are any additional comments

File: datafusion/expr/src/logical_plan/builder.rs [hunk 3]
```diff
@@ -991,6 +992,68 @@ pub fn table_scan(
...
+pub fn wrap_projection_for_join_if_necessary(
+    join_keys: &[Expr],
+    input: LogicalPlan,
+) -> Result<(LogicalPlan, Vec<Column>, bool)> {
+    let input_schema = input.schema();
+    let alias_join_keys: Vec<Expr> = join_keys
+        .iter()
+        .map(|key| {
+            // The display_name() of cast expression will ignore the cast info, and show the inner expression name.
+            // If we do not add alais, it will throw same field name error in the schema when adding projection.
```
Comment (liukun4515) [inline datafusion/expr/src/logical_plan/builder.rs:1005]: good comments

**Review by liukun4515 — APPROVED** [review]
Maybe we add sql test like in the follow up pr.
```
select *,t1.t1_id+10 from t1,t2 where t1.t1_id+10=t2.t2_id
```

Thanks @ygf11

**Comment by ygf11** [comment]
> Maybe we can add sql test in the follow up pr.

Thanks, I will add the test case in the follow up pr.

**Comment by alamb** [comment]
Thanks @ygf11  for the contribution and everyone for the reviews!
